### PR TITLE
[Traceable FSDP2] Do 1st iteration dry-run in eager mode without explicit training loop change (via LazyModule)

### PR DESCRIPTION
**Pull request by yf225**
Currently, a big blocker for Traceable FSDP2 adoption is the requirement to modify the training loop to run 1st iteration in eager mode to do the FSDP2 initialization, before torch.compile can kick in in 2nd iteration and beyond. This PR makes it so that the eager-mode FSDP2 initialization is run automatically, without the need for explicit training loop change.

Test commands:
- `pytest -rA test/distributed/_composable/fsdp/test_fully_shard_compile.py::TestFullyShardCompile::test_nested_fully_shard_fullgraph_backend_inductor`
- `pytest -rA test/distributed/_composable/fsdp/test_fully_shard_compile.py::TestFullyShardCompile::test_transformer_fullgraph_backend_inductor`

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #131721
* #131507

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @ezyang @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @amjames

### CODE REVIEW

**Comment by ezyang** [comment]
Hey @yf225, do we want to specifically hijack the LazyModule mechanism to do this? I would support a generalized warmup mechanism, e.g., on the torch.compile API itself, or even a dynamo config that affects everything, because I think it's quite common to want to run warmup before actual PT2; in fact, I have recently proposed that we should do this for some fixed N on all workloads to (1) potentially take advantage of async compilation and (2) to allow for a comparison between eager and compiled performance.

**Comment by yf225** [comment]
> Hey @yf225, do we want to specifically hijack the LazyModule mechanism to do this? I would support a generalized warmup mechanism, e.g., on the torch.compile API itself, or even a dynamo config that affects everything, because I think it's quite common to want to run warmup before actual PT2; in fact, I have recently proposed that we should do this for some fixed N on all workloads to (1) potentially take advantage of async compilation and (2) to allow for a comparison between eager and compiled performance.

@ezyang Sounds great - I will try a Dynamo config solution. I actually have a slightly more complex use case ("two-stage warmup") that I want to see if it's supportable:
1. FSDP2 needs eager 1st iteration to do state init (supportable by Dynamo warmup run)
2. FSDP2 then needs a `Dynamo(backend="eager")` 2nd iteration to check whether there is graph break (need new capability to support this).
3. And if there is graph break in 2nd iteration, in 3rd iteration we want to toggle `torch._dynamo.config.skip_fsdp_hooks` to True, so that all subsequent iterations will graph break on FSDP2 hooks to get best performance.

**Comment by yf225** [comment]
Dynamo warmup run feature: https://github.com/pytorch/pytorch/pull/131843
